TOY-32 다이어리 백업 및 복원 기능 제공

pby2017 opened the pull request:
코드 리뷰 부탁드립니다.
1. 백업 및 복원 기능 구현
2. 문제 해결
* 백업/복원 버튼 클릭으로 로그인했을 때 로그인 사용자 명 변경 반영
* 인터넷 끄고 백업/복원 시 앱 종료되는 문제 해결
* 서버에 데이터가 없는데 복원할 경우와 단말기에 데이터가 없는데 백업하는 경우 앱 종료되는 문제 해결

----------

람다 사용을 위해서 JDK 1.7 -> JDK 1.8로 변경한 사항이 있는데, 정상적으로 빌드는 되지만 람다 식에서 빨간색 밑줄(경고)이 그어지고 있습니다.
이후 코틀린으로 변환했을 때 해결될 것으로 보입니다.

pby2017 committed:
TOY-32 다이어리 백업 및 복원 기능 제공

- 백업, 복원 기능 구현

pby2017 committed:
TOY-32 피드백 반영

pby2017 committed:
TOY-32 develop merge 충돌 해결

pby2017 committed:
TOY-32 Merge branch 'develop' into TOY-32-backup-restore

# Conflicts:
#	app/build.gradle
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/BaseActivity.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/Constants.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/DiaryDetailActivity.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/DiaryDao.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/DiaryDatabaseCallback.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/LocalDiaryViewModel.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/gallery/GalleryActivity.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SettingsActivity.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SettingsFragment.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SignDialogFragment.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SignOutDialogFragment.java
#	app/src/main/res/values/strings.xml
#	app/src/main/res/xml/preference_settings.xml

pby2017 committed:
TOY-32 develop(with TOY-34) 충돌 해결

pby2017 committed:
TOY-32 gitignore 파일 제거

hyejin830 commented on .idea/codeStyles/Project.xml:
```diff
@@ -0,0 +1,29 @@
+<component name="ProjectCodeStyleConfiguration">
```
이 파일은 지우는게 더 좋을 것 같습니다~

pby2017 replied:
네,
피드백 반영했습니다.

hyejin830 commented:
빈 사진이 포함된 다이어리를 복원하면 테스트 이미지가 보여지고 있네요~ 확인 부탁드립니다!

hyejin830 commented:
빈 사진이 포함된 다이어리 내용을 편집하면 앱이 죽는 현상이 보이네요~

DiaryEditActivity.java의 line 198 코드를 이렇게 수정해주시면 될 것 같습니다!
```
 if (diary.getImage() != null) {
     photoFileName = diary.getImage();
     editPhotoImageButton.setImageURI(Uri.fromFile(new File(Constants.YOGIDIARY_PATH, photoFileName)));
}
```

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/sync/DiaryBackupRestore.kt:
```diff
@@ -0,0 +1,56 @@
+
+    fun executePostDiary(currentUser: FirebaseUser, diary: Diary): Flowable<ResponseBody> {
+        val builder = MultipartBody.Builder().apply {
+            setType(MEDIA_TYPE_MULTIPART!!)
```
MEDIA_TYPE_IMAGE와 다르게 !! 처리되어 있고 동반 객체에서 초기화를 해주는 것 같은데 null이 될 수도 있는 건가요?! 궁금해서 여쭤봅니다~

pby2017 replied:
`setType(MEDIA_TYPE_MULTIPART!!)` 의 경우에서 `setType()` 은 인자에 null값을 허용하고 있지 않고 있는 것으로 보입니다.
그리고 `MEDIA_TYPE_MULTIPART = MediaType.parse("multipart/form-data")`은  `parse()` 메서드가 `@nullable` 값을 리턴해줄 수 있습니다.
> MEDIA_TYPE_MULTIPART 는 null이 될 수 있지만 setType()에는 null을 전달하면 안되기 때문에 !!를 붙여줬습니다. 테스트 중 null이 된 적은 없습니다.

```kotlin
public Builder setType(MediaType type) {
    if (type == null) {
    throw new NullPointerException("type == null");
    }
    if (!type.type().equals("multipart")) {
    throw new IllegalArgumentException("multipart != " + type);
    }
    this.type = type;
    return this;
}
```

hyejin830 commented:
백업한 적이 없을 때, 복원을 하면 '백업 실패'라고 다이얼로그가 나오고 있습니다~ 텍스트를 바꾸면 더 좋을 것 같습니다~

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/Constants.java:
```diff
@@ -20,4 +24,16 @@
+    public static final String PREFERENCE_DIALOG_TAG = "android.support.v7.preference.PreferenceFragment.DIALOG";
+    public static final String COMPRESSED_FOLDER_NAME = "compressed";
+    public static final SimpleDateFormat serverDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
+    public static final FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
```
globalApplication에 작성하면 더 좋을 것 같습니다~

pby2017 replied:
네,
피드백 반영했습니다.

hyejin830 commented:
확인 완료하였습니다

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/Constants.java:
```diff
@@ -20,4 +24,16 @@
+    public static final String SIGN_DIALOG_KEY = "sign_dialog";
+    public static final String PREFERENCE_DIALOG_TAG = "android.support.v7.preference.PreferenceFragment.DIALOG";
+    public static final String COMPRESSED_FOLDER_NAME = "compressed";
+    public static final SimpleDateFormat serverDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
```
Contants 클래스에 있으면 안될거 같아요.

pby2017 replied:
네,
피드백 반영해서 SimpleDateFormat 객체를 사용하는 클래스의 private 멤버 변수로 정의했습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/Constants.java:
```diff
@@ -20,4 +24,16 @@
+    public static final String PREFERENCE_DIALOG_TAG = "android.support.v7.preference.PreferenceFragment.DIALOG";
+    public static final String COMPRESSED_FOLDER_NAME = "compressed";
+    public static final SimpleDateFormat serverDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
+    public static final FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
```
Contants 클래스에 있으면 안될거 같아요. Contants 의미랑 안어울리네요.

pby2017 replied:
네,
피드백 반영해서 GlobalApplication의 변수로 정의했습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/BaseActivity.kt:
```diff
@@ -42,6 +44,15 @@ open class BaseActivity : AppCompatActivity() {
         progressBar.visibility = View.GONE
     }
 
+    fun buildAlertDialog(context: Context, message: String): AlertDialog.Builder {
```
positve button listener랑 negative button text, listener를 default 값을 두고 함수를 정의하는 편이 좋겠어요.

pby2017 replied:
네, 
피드백 반영해서 button 텍스트 및 listener 인자 추가와 default 값을 두도록 수정했습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/DiaryDetailActivity.java:
```diff
@@ -116,7 +116,9 @@ private void setViewData(int diaryIndex) {
                         setDateAndTime(diary.getDate());
-                        detailPhotoImageView.setImageURI(Uri.fromFile(new File(Constants.YOGIDIARY_PATH, diary.getImage())));
+                        if (diary.getImage() != null) {
+                            detailPhotoImageView.setImageURI(Uri.fromFile(new File(Constants.YOGIDIARY_PATH, diary.getImage())));
```
Glide로 이미지 로드하면 안되나요?

pby2017 replied:
가능합니다.
null 체크 코드를 제거하고 Glide로 수정했습니다.

이미지 뷰 코드를 Glide 코드로 교체하는 작업이 따로 계획되어 있어서 
코드를 바꾸지 않고 null 체크 코드만 추가했었습니다.

bonchangkoo reviewed (commented):
코드 확인했습니다.

pby2017 commented:
> 백업한 적이 없을 때, 복원을 하면 '백업 실패'라고 다이얼로그가 나오고 있습니다~ 텍스트를 바꾸면 더 좋을 것 같습니다~

```
// 예외 종류
`NO_DATA` 상황 : 이전에 백업했었는데 서버 데이터가 지워진 후(clear) 복원을 시도할 때 크기가 0인 Array<Diary>가 반환된 상태
`NOT ARRAY BUT OBJECT` 상황 : 회원가입 후 백업 이력 없이 복원을 시도할 때 Array<Diary>가 아닌 Object가 반환된 상태
```

@hyejin830 님,
`NOT ARRAY BUT OBJECT` 상황이 조건 분기문에서 따로 구분하지 않아 실패 메시지로 표시되었는데
피드백 반영해서 `NO_DATA` 상황이랑 같은 메시지를 보여주도록 했습니다. (메시지 = `서버에 저장한 다이어리가 없습니다.`)

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java:
```diff
@@ -0,0 +1,322 @@
+    }
+
+    private boolean writeResponseBodyToDisk(String imageFileName, ResponseBody body) {
+        try {
```
~~~Java
// 아래 로직들은 restoredImageFile이 없으면 의미 없는 로직이니 맨 앞에서 이렇게 파일 존재유뮤 체크해주고 가면 좋을것 같습니다.
File restoredImageFile = new File(Constants.YOGIDIARY_PATH, imageFileName);
if(! restoredImageFile.exist(){
    return false
}

// try catch resource :Java7 이러면 302번 라인의 finally를 대체할수 있겠네요
try(inputStream = body.byteStream();
      outputStream = new FileOutputStream(restoredImageFile)
){
.....

~~~

pby2017 replied:
네,
피드백 반영해서 바깥 try 블록을 제거하고,
try catch resource 방식으로 수정했습니다.

pby2017 replied:
```java
 if(! restoredImageFile.exist()){  
     return false  
 }
```

위 코드에서 exist()의 경우 단말기에 저장되어 있는지 확인하는 코드 같아 보입니다.
해당 코드를 추가하고 실행했을 때, 아래 코드는 서버에서 이미지 String 데이터를 가져온 직후이며 아직 파일형태로 저장한 상태가 아니라 `restoredImageFile.exist() = false` 값이 나오는 것 같습니다.

```java
File restoredImageFile = new File(Constants.YOGIDIARY_PATH, imageFileName);
```

따라서 해당 조건문은 현재 반영하지 않았습니다.

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java:
```diff
@@ -0,0 +1,322 @@
+                inputStream = body.byteStream();
+                outputStream = new FileOutputStream(restoredImageFile);
+
+                while (true) {
```
while(true) 와 같이 무한루프를 코드에 넣는 것은 위험합니다  가능한 피해주세요
또한 Nested try catch 도 안티 패턴입니다 
~~~Java
while(int read =inputStream.read(input, 0, input.length)) != -1){
     outputStream.write(fileReader, 0, read);
     fileSizeDownloaded += read;
     
     Log.d(TAG, "file download: " + fileSizeDownloaded + " of " + fileSize);
    .....
~~~

pby2017 replied:
네,
피드백 참고하겠습니다. 반영해서 아래와 같이 수정했습니다.
```java
int read = inputStream.read(input, 0, input.length)
while(read != -1) {
    ...
    read = inputStream.read(input, 0, input.length)
}
```

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java:
```diff
@@ -0,0 +1,322 @@
+                context.sendBroadcast((new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(restoredImageFile))));
+
+                return true;
+            } catch (IOException e) {
```
Nested try catch일 이유가 없었네요  서로 같은 IOException 을 잡고 같은 return false; 이니까요

pby2017 replied:
네,
말씀 주신대로 중복되는 체크 예외 코드였습니다.
피드백 반영했습니다.

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SettingsActivity.java:
```diff
@@ -1,31 +1,27 @@
 
-        // TODO : global application에 구현하면 어떨지 고민
-        firebaseAuth = FirebaseAuth.getInstance();
+        ImageButton backspaceButton = findViewById(R.id.btn_backspace);
```
[질문질문!] 이건 버터 나이프로 처리 안되나요?

pby2017 replied:
제가 어제 잘못 말씀드렸던 것 같습니다.

먼저, 질문 주신 의도가 findViewById를 없앨 수 있는 방법으로 버터 나이프를 사용하기 때문이라고 이해했습니다.

어제 `databinding` 이라는 것으로 그것을 대신한다고 말씀드렸었는데 잘못 말씀드렸던 것 같습니다.

대신 코틀린 언어를 사용할 경우
`kotlin android extension`을 제공해주는데 
이것이 findViewById를 생략하고 바로 id 값으로 View 객체에 접근할 수 있도록 지원해줍니다.

따라서 java에서 kotlin으로 변환하면 해결될 것으로 보입니다.

피드백 주신 덕분에 다시 한 번 학습했습니다.

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/Constants.java:
```diff
@@ -20,4 +24,16 @@
     // PATH
     public static final File YOGIDIARY_PATH = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES), "YogiDiary");
```
자세히는 모르지만  
ROOTPATH를 선언해두기 위함이라면 문자열로 선언되있는게 더 좋을것 같아요 
String 이 아닌 File인 이유는 어떤 이유가 있는건가요?

pby2017 replied:
해당 `File 변수를 File 타입으로 사용`하는 곳은 사진을 저장하기 위한 폴더의 존재를 확인하고 없다면 생성할 때 사용하고 있습니다. [코드 링크](https://github.com/bonchangkoo/RookiePhotoApp/blob/develop/app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/DiaryEditActivity.java#L409)

나머지 사용 용도는 `위 링크의 417번 라인`과 같이
`File 변수의 Path String 값을 반환`해서 다른 String과 연결(concat) 한 뒤 새로운 File 객체로 생성할 때 사용하고 있습니다.

혹시나 제가 모르는 사용 용도가 있을 수 있으니 피드백 주신 부분 관련해서 얘기해보고 반영해보겠습니다.

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java:
```diff
@@ -0,0 +1,322 @@
+
+    private Compressor imageCompressor;
+
+    private CompositeDisposable compositeDisposable = new CompositeDisposable();
```
이 객체와 DiaryBackupRestore 는 onCreate()에서 인스턴스를 생성할수는 없나요?

안드로이드 구현체?면 생명주기를 벗어나서 혼자서 따로 인스턴스가 생성되는 경우는 없는 걸로 알고있어서요(뇌피셜)...

pby2017 replied:
onCreate() 에서 생성할 수 있습니다.

클래스 바로 아래 인스턴스 변수로 선언하고 바로 초기화해서 인스턴스를 생성하는 것에 대해서 의심을 가져본 적은 없지만,

클래스가 newInstance() 메서드에서 인스턴스화되면서 Java의 클래스를 생성하는 것과 같이 
변수의 인스턴스가 생성되는 것으로 이해했습니다.

대신 인스턴스를 생성할 때 전달할 파라미터로 안드로이드의 context나 this가 필요한 경우는
라이프 사이클 메서드 안에서 인스턴스를 생성해줘야 합니다.

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java:
```diff
@@ -0,0 +1,322 @@
+                    dismiss();
+                }
+                setLoadingDialog();
+                String preferenceKey = getArguments().getString(Constants.PREFERENCE_KEY);
```
105번째와 동일한 로직이네요 ...

pby2017 replied:
네,
해당 라인의 중복은 줄일 수 있도록 수정했습니다.

80번째 라인 `onBindDialogView()` 과
103번째 라인 `onPrepareDialogBuilder()` 에도 동일한 로직이 존재하는데
이 부분은 리팩토링을 하면서 둘 중 하나만 사용하도록 수정하려고 합니다.

KimSoungRyoul reviewed (commented):
제 피드백보다는 본창님의 피드백을 좀더 우선 해주세요
안드로이드는 잘몰라서요

pby2017 commented:
> 빈 사진이 포함된 다이어리를 복원하면 테스트 이미지가 보여지고 있네요~ 확인 부탁드립니다!

@hyejin830 님, 
피드백 반영해서 해당 부분(detail 화면)과 메인 부분(main 화면)에서 보여지는 이미지를 모두 공백으로 처리했습니다. 해당 영역을 어떤 것으로 채울지 고민하는 것이 길어질 것 같아서 공백으로 두는 것을 선택했습니다.

pby2017 commented:
> 빈 사진이 포함된 다이어리 내용을 편집하면 앱이 죽는 현상이 보이네요~
> 
> DiaryEditActivity.java의 line 198 코드를 이렇게 수정해주시면 될 것 같습니다!
> 
> ```
>  if (diary.getImage() != null) {
>      photoFileName = diary.getImage();
>      editPhotoImageButton.setImageURI(Uri.fromFile(new File(Constants.YOGIDIARY_PATH, photoFileName)));
> }
> ```

@hyejin830 님,
피드백 반영하면서
`new File()`의 파라미터로 null 값이 전달될 때 문제가 발생하는 것을 확인했습니다.
`new File()`로 생성하지 않고 String 경로 값을 생성하도록 수정했습니다.

그리고 null 체크를 Glide에게 위임하는 형태로 바꿨습니다.

pby2017 committed:
TOY-32 피드백 반영

pby2017 committed:
TOY-32 충돌 해결, LocalDiaryViewModel의 nullable 인자의 타입에 '?' 키워드 추가

Merge branch 'develop' into TOY-32-backup-restore

# Conflicts:
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/GlobalApplication.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/DiaryEditActivity.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/DiaryDao.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/DiaryDatabaseCallback.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/LocalDiaryViewModel.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/edit/EditPhotoActivity.kt

pby2017 commented:
@bonchangkoo 님, @hyejin830 님, @KimSoungRyoul 님,

피드백 반영 완료했습니다.
확인 부탁드립니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/DiaryEditActivity.java:
```diff
@@ -156,12 +160,12 @@ public void onClick(DialogInterface dialog, int which) {
     @Override
     public void onBackPressed() {
         AlertDialog.Builder builder = new AlertDialog.Builder(DiaryEditActivity.this);
```
여기서는 왜 AlertDialog.Builder 쓰는거에요? base에 만들어 뒀는뎀

pby2017 replied:
놓친 부분이었습니다.
피드백 반영해서 AlertDialog.Builder를 사용하면서 BaseActivity의 메서드를 사용할 수 있는 코드는 전부 수정했습니다.

bonchangkoo reviewed (commented):
코멘트 확인 요청 드려요~

pby2017 committed:
TOY-32 피드백 반영 (AlertDialog.Builder)

bonchangkoo reviewed (approved):
확인 했습니다. 제가 바로 머지할게요.
